docs/contrib/sequences.rst

```diff
@@ -0,0 +1,79 @@
+==============
+Lazy sequences
+==============
+
+.. versionadded:: 0.12.0
+
+Sequences module contains few macros for declaring sequences that are evaluated
+only as much as the client code requests elements. Compared to generators they
+allow accessing same element multiple times. Since they cache calculated
+values, they aren't suited for infinite sequences. However, the implementation
+allows recursive definition of sequences, without resulting recursive
+computation.
+
+To use these macros you need to require them and import other types like:
+
+.. code-block:: hy
+
+   (require [hy.contrib.sequences [defseq seq]])
+   (import [hy.contrib.sequences [Sequence end-sequence]])
+
+Simplest sequence can be defined as: ``(seq [n] n)``. This defines a sequence
+that starts as: ``[0 1 2 3 ...]`` and continues forever. In order to define
+a finite sequence, ``end-sequence`` needs to be called to signal end of
+the sequence:
+
+.. code-block:: hy
+
+   (seq [n]
+        "sequence of 5 integers"
+        (cond [(< n 5) n]
+              [true (end-sequence)]))
+
+This creates following sequence: ``[0 1 2 3 4]``. For such a sequence ``len``
+returns amount of items in sequence and negative indexing is suported. Because
+both of thse require evaluating whole sequence, calling such a function would
+take forever (or at least until available memory has been exhausted).
+
+Sequence can be defined recursively. Canonical example of fibonacci numbers
```

Comment:
**Sequences** or **The sequence** or **A sequence**

**The/a canonical example**